### DISCUSSION TITLE
Overlay for non-modal dialog?

### DISCUSSION THREAD

**Discussion by vnugent** [radix-ui/primitives#1671] [Help]
I notice when Dialog.modal = false ([doc](https://www.radix-ui.com/docs/primitives/components/dialog#root)) the overlay isn't rendered. 

Is it possible to change this behavior?  that is showing the overlay when modal = false

In our Dialog component there's a custom Algolia autocomplete search widget which acts like its own modal and portal, mounting itself to `body` on touch / mouseclick.   The issue is Radix Dialog in modal mode (default) disallows interactions out of the dialog preventing user interacting with the auto-complete search box once it's activated.

We're forced to use Dialog in non-modal mode.  But we need to be able to show the overlay.   Any suggestions are greatly appreciated.

**Comment by benoitgrelard** [comment]
Hey @vnugent,

The reason there is no overlay for non-modal mode is because it doesn't make sense to have one as non-modal means you can interact with what's outside the dialog.

Rather it would be good to understand your actual issue (the one forcing you to reluctantly opt-in for non-modal mode) and see if we can provide solutions there.

**Comment by vnugent** [comment]
Our wizard component is wrapped in a Radix modal.   Since we use Algolia search component in [detached mode](https://www.algolia.com/doc/ui-libraries/autocomplete/core-concepts/detached-mode/), the search box once triggered, is mounted to a body portal, which Radix correctly prevents interaction because it's outside of the wizard.

Is there a workaound that allows us to be able to use the wizard (Radix Dialog) in modal mode?

In my view the overlay is simply an optional visual cue to minimize distraction in order to direct attention to the dialog.  Whether or not you can interact outside of the dialog should be controlled by the modal flag.

<img width="695" alt="Screen Shot 2022-09-20 at 3 50 33 PM" src="https://user-images.githubusercontent.com/3805254/191278781-b43097b7-ff55-4d5e-8daa-bd74d51ae050.png">

**Reply by benoitgrelard** [reply]
> Is there a workaound that allows us to be able to use the wizard (Radix Dialog) in modal mode?

I guess here you end up with a modal within a modal, which is a known dark pattern in terms of UX.
Have you looked at building your own custom UI for the algolia search? Or not using detached mode?

> In my view the overlay is simply an optional visual cue to minimize distraction in order to direct attention to the dialog. Whether or not you can interact outside of the dialog should be controlled by the modal flag.

The overlay is an affordance that what is behind is indeed blocked (and at its core, even without javascript, by default it would be blocked). Hence why it's inherently linked to modality, if the user can in fact interact with what's behind, then no overlay should be shown.

Technically it is very easy to have your own overlay for the non-modal dialog, BUT I don't think that is the solution here.

If you can build a quick codesandbox to show your issue, I can take a look and see if I can find simpler workarounds otherwise.

**Reply by vnugent** [reply]
> Have you looked at building your own custom UI for the algolia search? Or not using detached mode?

The one in the screenshot is actually a custom built component based on Algolia autocomplete.  Being a small team we decided to use detached mode for both mobile and desktop so that there is less code and css to maintain.

> I guess here you end up with a modal within a modal, which is a known dark pattern in terms of UX.

In general I agree with you.  However, it's more common than you think when the 2nd modal is very small like an alert.  See screenshot of IG photo upload in desktop mode below.  Also for web mobile sometimes we want to emulate "native input" mode by having a large modal (which what Algolia detached mode is for)

Anyhow we will see if we can add support non-detached mode to our search component.  Thanks for your help.
<img width="400" alt="Screen Shot 2022-09-20 at 4 51 16 PM" src="https://user-images.githubusercontent.com/3805254/191295156-99a9b4fe-88ba-4a37-9202-a2e56b255f7f.png">

**Comment by cill-i-am** [comment]
Sorry to bring up an old thread @benoitgrelard, but we are having the same issue with our intercom widget. It is unclickable unless modal={false}

**Reply by benoitgrelard** [reply]
What is the issue?

**Reply by cill-i-am** [reply]
The default dialog behaviour disallows clicking the intercom widget:
https://user-images.githubusercontent.com/8224386/221875269-8fdedbf7-08b3-4cfb-94d3-9f3f4a0da9c6.mov

but setting modal={false} on the dialog root (and providing our own overlay) works as expected:
https://user-images.githubusercontent.com/8224386/221875831-e668c3ce-32c0-4967-abbf-5db00ada36b5.mov

It would be nice if there was some way to tell the overlay to trigger even if the root has modal={false}. It would save us having to roll a custom overlay. It's a small thing though

**Reply by benoitgrelard** [reply]
A non-modal dialog with an overlay doesn't make sense though, as modal means nothing can be interacted outside as this is considered at the forefront.

Ideally, depth-wise, this means that your intercom widget should be below the overlay.

**Reply by cill-i-am** [reply]
I agree, but a lot of our dialogs contain forms, and the widget has to be clickable for customer support reasons

**Reply by benoitgrelard** [reply]
There is a workaround, but it would only be accessible to pointer users: set `pointer-events: auto` on your intercom widget to ensure it is still interactive. That would make it work for pointer, but you still wouldn't be able to tab to it, making it inaccessible for keyboard users. Perhaps a tradeoff you would be willing to accept.

**Reply by cill-i-am** [reply]
Hmm that could be useful. Thanks for the help 🙏
